File: cli/src/api/bitbucket.rs
use serde::Deserialize;
use reqwest::blocking::Client;

#[derive(Debug, Deserialize)]
pub struct BitbucketContribution {
    pub date: String,
    pub count: u32,
}

#[derive(Debug)]
pub struct BitbucketIssue {
    pub title: String,
    pub url: String,
}

pub fn fetch_bitbucket_contributions(token: &str) -> Vec<BitbucketContribution> {
    let client = Client::new();
    let url = "https://api.bitbucket.org/2.0/repositories?role=member";
    let resp = client.get(url)
        .bearer_auth(token)
        .send();
    if resp.is_err() {
        return vec![];
    }
    let repos: serde_json::Value = resp.unwrap().json().unwrap_or_default();
    let mut contributions = Vec::new();
    if let Some(values) = repos["values"].as_array() {
        for repo in values.iter().take(5) {
            let owner = repo["owner"]["username"].as_str().unwrap_or("");
            let slug = repo["slug"].as_str().unwrap_or("");
            let url = format!("https://api.bitbucket.org/2.0/repositories/{}/{}/commits", owner, slug);
            let commits_resp = client.get(&url)
                .bearer_auth(token)
                .send();
            if let Ok(commits) = commits_resp {
                let commits_json: serde_json::Value = commits.json().unwrap_or_default();
                let count = commits_json["values"].as_array().map(|v| v.len()).unwrap_or(0);
                contributions.push(BitbucketContribution {
                    date: "recent".to_string(),
                    count: count as u32,
                });
            }
        }
    }
    contributions
}

pub fn fetch_bitbucket_issues(token: &str) -> Vec<BitbucketIssue> {
    let client = Client::new();
    let url = "https://api.bitbucket.org/2.0/repositories?role=member";
    let resp = client.get(url)
        .bearer_auth(token)
        .send();
    if resp.is_err() {
        return vec![];
    }
    let repos: serde_json::Value = resp.unwrap().json().unwrap_or_default();
    let mut issues = Vec::new();
    if let Some(values) = repos["values"].as_array() {
        for repo in values.iter().take(5) {
            let owner = repo["owner"]["username"].as_str().unwrap_or("");
            let slug = repo["slug"].as_str().unwrap_or("");
            let url = format!("https://api.bitbucket.org/2.0/repositories/{}/{}/issues?q=state=\"new\" AND kind=\"bug\"", owner, slug);
            let issues_resp = client.get(&url)
                .bearer_auth(token)
                .send();
            if let Ok(issues_json) = issues_resp {
                let issues_val: serde_json::Value = issues_json.json().unwrap_or_default();
                if let Some(arr) = issues_val["values"].as_array() {
                    for issue in arr {
                        issues.push(BitbucketIssue {
                            title: issue["title"].as_str().unwrap_or("").to_string(),
                            url: issue["links"]["html"]["href"].as_str().unwrap_or("").to_string(),
                        });
                    }
                }
            }
        }
    }
    issues
}

File: web/uku-backend/services/bitbucket.js
const axios = require('axios');

const BITBUCKET_API_BASE = 'https://api.bitbucket.org/2.0';
const BITBUCKET_TOKEN = process.env.BITBUCKET_TOKEN || '';

async function fetchBitbucketContributions() {
  try {
    const reposResp = await axios.get(`${BITBUCKET_API_BASE}/repositories?role=member`, {
      headers: {
        'Authorization': `Bearer ${BITBUCKET_TOKEN}`,
      },
    });
    const repos = reposResp.data.values.slice(0, 5);
    const contributions = [];
    for (const repo of repos) {
      const commitsResp = await axios.get(`${BITBUCKET_API_BASE}/repositories/${repo.owner.username}/${repo.slug}/commits`, {
        headers: {
          'Authorization': `Bearer ${BITBUCKET_TOKEN}`,
        },
      });
      contributions.push({
        platform: 'Bitbucket',
        date: 'recent',
        count: commitsResp.data.values.length,
      });
    }
    return contributions;
  } catch (error) {
    console.error('Bitbucket fetch error:', error.message);
    return [];
  }
}

async function fetchBitbucketIssues() {
  try {
    const reposResp = await axios.get(`${BITBUCKET_API_BASE}/repositories?role=member`, {
      headers: {
        'Authorization': `Bearer ${BITBUCKET_TOKEN}`,
      },
    });
    const repos = reposResp.data.values.slice(0, 5);
    const issues = [];
    for (const repo of repos) {
      const issuesResp = await axios.get(`${BITBUCKET_API_BASE}/repositories/${repo.owner.username}/${repo.slug}/issues?q=state="new" AND kind="bug"`, {
        headers: {
          'Authorization': `Bearer ${BITBUCKET_TOKEN}`,
        },
      });
      for (const issue of issuesResp.data.values) {
        issues.push({
          platform: 'Bitbucket',
          title: issue.title || '',
          url: issue.links.html.href || '',
        });
      }
    }
    return issues;
  } catch (error) {
    console.error('Bitbucket issues fetch error:', error.message);
    return [];
  }
}

module.exports = { fetchBitbucketContributions, fetchBitbucketIssues };